docs/local-development.md

```diff
@@ -285,6 +285,105 @@ function onClick(event) {
   - [Type Inconsistencies](./flow.md#type-inconsistencies)
 
+### Reducers
+
+Our reducers are where we store the debugger state. We try to follow Redux best
+practices, but have added our own flavor as well with the help of Flow and Immutable
+
+#### Flow
+
+We type our stores so that we can document the shape of the data and guarantee
+the data coming in and out is well formed.
+
+Lets look at the [expressions] reducer and see how it is typed:
+
+```js
+type ExpressionState = {
+  expressions: List<Expression>
+};
+
+export const State = makeRecord(
+  ({
+    expressions: List()
+  }: ExpressionState)
+);
+
+function update(
+  state: Record<ExpressionState> = State(),
+  action: Action
+): Record<ExpressionState> {
+  // ...
+}
+
+type OuterState = { expressions: Record<ExpressionState> };
+
+export function getExpressions(state: OuterState, input: string) {
+  return getExpressions(state).find(exp => exp.input == input);
+}
+```
+
+The `ExpressionState` documents the reducers fields. We use it in three places:
+
+1. `State` - an Immutable expression state record
+2. `update` - the reducer function which receives the expression state record
+3. `OuterState` - a local type representing the application state passed into selectors
+
+#### Immutable
+
+We try to wrap our state in Immutable records when we can for two reasons.
+First it means that the state can only be modified in the reducers.
+Second, it helps our connected components avoid unecessary renders.
+
+Connect will trigger a re-render when it sees new state, even if it has not changed.
+Immutable, will creates new objects if and only if the data changes,
+which means our components only render when it's approriate.
+
+The one situation where we do not use immutable is when it is to slow to do so.
+We avoid wrapping our pause state in immutable, because it takes to long to serialize the data.
+
+Lets take a look at the Expressions reducer to see how we use Immutable.
+
+```js
+type ExpressionState = {
+  expressions: List<Expression>
+};
+
+export const State = makeRecord(
+  ({
+    expressions: List()
+  }: ExpressionState)
+);
+
+function update(
+  state: Record<ExpressionState> = State(),
+  action: Action
+): Record<ExpressionState> {
+  case "DELETE_EXPRESSION":
+    return deleteExpression(state, action.input);
+  // ...
+}
+
+type OuterState = { expressions: Record<ExpressionState> };
+
+function deleteExpression(state: State, input: string) {
+  const index = getExpressions({ expressions: state }).findKey(
+    e => e.input == input
+  );
+  return state.deleteIn(["expressions", index]);
+}
+```
+
+The first thing to notice is that the `Expression` is an Immutable list.
+We document that here `List<Expression>` and here `expressions: List()`.
+
+The second thing to note is that we use the Immutable api to update the state.
+We do that in `deleteExpression` here `state.deleteIn`. There [docs][immutable-docs] are really helpful.
```

Comment:
we will need to add "[immutable-docs]: https://facebook.github.io/immutable-js/docs/#/"